contrib/babelfishpg_tsql/src/pltsql_coerce.c

```diff
@@ -888,6 +921,227 @@ run_tsql_best_match_heuristics(int nargs, Oid *input_typeids, FuncCandidateList
+							errmsg("Failed to find common utility plugin.")));
+
+				special_func->valid_arg_types[i].valid_types_oid[j] = (*common_utility_plugin_ptr->get_tsql_datatype_oid)(special_func->valid_arg_types[i].valid_types[j]);
+			}
+
+			validTypId = special_func->valid_arg_types[i].valid_types_oid[j];
+
+			if (inputTypId == validTypId)
+			{
+				type_match = true;
+				break;
+			}
+		}
+		if (!type_match)
+		{
+			ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_FUNCTION),
+				 errmsg("Argument data type %s is invalid for argument %d of %s function.", 
+				 		 format_type_be(inputTypId), i+1, special_func->formatted_funcname)));
+		}
+	}
+
+	return true;
+}
+
+/*
+ * tsql_func_select_candidate_for_special_func()
+ *
+ * For functions present in special function list, and try to find best candidate 
+ * based on matching return type. Also throw error in case of invalid argument data type.
+ */
+static FuncCandidateList
+tsql_func_select_candidate_for_special_func(List *names, int nargs, Oid *input_typeids, FuncCandidateList candidates)
+{
+	FuncCandidateList			current_candidate, best_candidate;
+	Oid 						expr_result_type;
+	char					   *proc_nsname;
+	char					   *proc_name;
+	bool						is_func_validated;
+	int							ncandidates;
+	Oid							retType;
+
+	DeconstructQualifiedName(names, &proc_nsname, &proc_name);
+
+	is_func_validated = validate_special_function(proc_nsname, proc_name, nargs, input_typeids);
+
+	/* Return NULL if function is not a special function */
+	if (!is_func_validated)
+		return NULL;
+
+	/* if common_utility_plugin_ptr is not initialised */
+	if (common_utility_plugin_ptr == NULL)
+		ereport(ERROR,
+				(errcode(ERRCODE_INTERNAL_ERROR),
+					errmsg("Failed to find common utility plugin.")));
+
+	/* function based logic to decide return type */
+	expr_result_type = InvalidOid;
+	if (strlen(proc_name) == 4 && strncmp(proc_name,"trim", 4) == 0)
+	{
+		if ((*common_utility_plugin_ptr->is_tsql_nvarchar_datatype)(input_typeids[1])
+			|| (*common_utility_plugin_ptr->is_tsql_nchar_datatype)(input_typeids[1]))
+		{
+			expr_result_type = (*common_utility_plugin_ptr->lookup_tsql_datatype_oid) ("nvarchar");	
+		}
+		else if ((*common_utility_plugin_ptr->is_tsql_varchar_datatype)(input_typeids[1])
+				|| (*common_utility_plugin_ptr->is_tsql_bpchar_datatype)(input_typeids[1])
+				|| input_typeids[1] == UNKNOWNOID)
+		{
+			expr_result_type = get_sys_varcharoid();
+		}
+	}
+
+	if (!OidIsValid(expr_result_type))
+		return NULL;
+
+	/* Get the candidate with matching return type */
+	ncandidates = 0;
+	best_candidate = NULL;
+	for (current_candidate = candidates;
+			current_candidate != NULL;
+			current_candidate = current_candidate->next)
+	{
+		retType = get_func_rettype(current_candidate->oid);
+		if (expr_result_type == retType)
+		{
+			best_candidate = current_candidate;
+			ncandidates++;
+		}
+	}
+
+	/* Only one definition should exists per return type for special function */
+	Assert(ncandidates == 1);
```

Comment:
This assert will not work in release builds, also seems redundant since you are already trimming the list below and only passing the first candidate in the list. If we really want to restrict we should throw an error

Reply:
This assert is added for developers to understand that they should only create one definition per return type for a special function. But in case if some developer creates multiple definitions with same return type for a special function and due to lack of enough tests, developer did not hit at this check then later when user uses this special function, we don't want them to get an unexpected internal error hence used assert here and choose the first candidate. This might cause unexpected output but it is better than throwing unexpected internal error to customer.

cc: @Deepesh125

Reply:
Thrown error instead of assert.